src/auth/impersonated.ts

```diff
@@ -0,0 +1,165 @@
+   * impersonated credentials
+   * @param targetPrincipal the service account to impersonate.
+   * @param delegates the chained list of delegates required to grant the
+   *  final access_token. If set, the sequence of identities must have
+   * "Service Account Token Creator" capability granted to the preceding
+   * identity. For example, if set to [serviceAccountB, serviceAccountC],
+   * the sourceCredential must have the Token Creator role on serviceAccountB.
+   * serviceAccountB must have the Token Creator on serviceAccountC.
+   * Finally, C must have Token Creator on target_principal.
+   * If left unset, sourceCredential must have that role on targetPrincipal.
+   * @param targetScopes scopes to request during the authorization grant.
+   * @param lifetime number of seconds the delegated credential should be
+   * valid for up to 3600 seconds by default, or 43,200 seconds by
+   * extending the token's lifetime, see:
+   * https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials#sa-credentials-oauth
+   * @param endpoint api endpoint override.
+   */
+  constructor(options: ImpersonatedOptions = {}) {
+    super(options);
+    this.credentials = {
+      expiry_date: 1,
+      refresh_token: 'impersonated-placeholder',
+    };
+    this.sourceClient = options.sourceClient || new OAuth2Client();
+    this.targetPrincipal = options.targetPrincipal || '';
+    this.delegates = options.delegates || [];
+    this.targetScopes = options.targetScopes || [];
+    this.lifetime = options.lifetime || 3600;
+    this.endpoint = options.endpoint || 'https://iamcredentials.googleapis.com';
+  }
+
+  /**
+   * Refreshes the access token.
+   * @param refreshToken Unused parameter
+   */
+  protected async refreshToken(
+    refreshToken?: string | null
+  ): Promise<GetTokenResponse> {
+    if (this.isTokenExpiring()) {
+      try {
+        await this.sourceClient.getAccessToken();
+        const name = 'projects/-/serviceAccounts/' + this.targetPrincipal;
+        const u = `${this.endpoint}/v1/${name}:generateAccessToken`;
+        const body = {
+          delegates: this.delegates,
+          scope: this.targetScopes,
+          lifetime: this.lifetime + 's',
+        };
+        const res = await this.sourceClient.request({
+          url: u,
+          data: body,
+          method: 'POST',
+        });
+        const tokenResponse = res.data as TokenResponse;
+        this.credentials.access_token = tokenResponse.accessToken;
+        this.credentials.expiry_date =
+          Date.parse(tokenResponse.expireTime) / 1000;
+        return {
+          tokens: this.credentials,
+          res,
+        };
+      } catch (error) {
+        const status = error?.response?.data?.error?.status;
+        const message = error?.response?.data?.error?.message;
+        if (status && message) {
+          throw new Error(`${status}: unable to impersonate: ${message}`);
```

Comment:
Both of these will swallow the original exception, and hide the call stack.  Instead of throwing a new error, I would suggest crafting this string, and then prepending it to the message of the existing error:

```js
error.message = `${status}: unable to impersonate: ${message} \n\n ${err.message}`
throw error;
```

Here and in the else statement.